docs/spec/consensus/fork-detection.md

```diff
@@ -0,0 +1,93 @@
+conflicting commits *C1* and *C2*, where *C1* is a commit from the main chain. The detection procedure works like this:  
+
+1) if there are processes from the set *signers(C2)* that are not part of *C1.valset*, they misbehaved as they are signing 
+   protocol messages in heights they are not validators. In this case commit *C2* is a self-contained evidence of misbehavior 
+   for those processes that can be simply verified by every honest validators without additional information.  
+   For processes from the set *signers(C2)* that are part of *C1.valset* we need additional checks:
+
+2) if *C1.round == C2.round*, and some processes signed different precommit messages in both commits, then it is an 
+   equivocation misbehavior. Similarly as above, commits *C1* and *C2* are an evidence of misbehavior. Note that in this case
+   there might be additional misbehavior, but its detection require more complex detection procedure that we explain next.  
+
+3) if *C1.round != C2.round* we need to run full detection procedure. We assume (in the first version) that the full detection
+   procedure is executed by a centralized, trusted component called monitor that will be processing proof of forks and 
+   identifying faulty processes (and optionally generating evidences of misbehavior). Monitor is an honest full node that runs 
+   on the main chain. Monitor is triggered by receiving two conflicting commits where *C1.round != C2.round*. The procedure starts
+   by declaring processes that are in the set *C1.valset* as suspected. This can be for example done by posting a special transaction
+   on the main chain. After suspected processes are declared, they are obliged to submit vote sets (prevote and precommit messages) for 
+   the given height for the subset of rounds *[C1.round, C2.round]* to the monitor process within *MONITOR_RESPONSE_PERIOD*.
+   Validators that do not submit its vote sets within this period are considered faulty (Q: how can we create verifiable
+   proof of absence of vote sets?). 
+
+   Once monitor receives vote sets from all validators or after *MONITOR_RESPONSE_PERIOD* is over, it executes the following
+   protocol:
+
+   1) Preprocessing step: the received messages from all vote sets are added to the appropriate vote sets of the senders. 
+      More precisely, if in the vote set of process *p1* there is a message *m* sent by process *p2*, we will add this message to
+      the vote set of the process *p2*. In essence, during this step we are computing what we call *complete vote sets*: union of 
+      vote sets sent by some process *p* together with messages sent by *p* that are found in other vote sets.
+   
+   2) Analysis step: we analyze vote sets of all processes independently and determine whether they are correct or faulty. Analysis 
+      of the vote set works as follows:
+      
+      - if more than one PREVOTE/PRECOMMIT message is sent in a round - faulty process
```

Comment:
1) what are the incentives for faulty process to send data of it's misbehavior? 
2) faulty process can try to cheat by sending "correct" vote sets, no?

Reply:
There are no incentives for doing it. Faulty process will be punished no matter whether it sends its data or not, in case correct processes do its work. Faulty process can try to cheat by sending correct votes, but as we are building their vote set based on what they send us and what other processes send us about them, this will potentially just create even more problems for them. So once faulty process does something bad, it's game over for them in case correct processes do they work correctly.